File: 12_bermudan_swaption_lsm_pricing/cpp/LSMBermudanPricer.cpp
#include "LSMBermudanPricer.hpp"
#include "RegressionBasis.hpp"

#include <cmath>
#include <iostream>
#include <iomanip>
#include <algorithm>
#include <stdexcept>

/*
    Constructor.
*/

LSMBermudanPricer::LSMBermudanPricer(
    const HullWhiteModel* model_,
    const BermudanSwaption* instrument_,
    const std::vector<double>& times_,
    const std::vector<std::vector<double>>& ratePaths_,
    const std::string& basisType_,
    int basisDegree_
){

    model = model_;
    instrument = instrument_;

    times = times_;
    ratePaths = ratePaths_;

    basisType = basisType_;
    basisDegree = basisDegree_;

    for(double t : instrument->exerciseDates){

        exerciseIndices.push_back(
            nearestTimeIndex(
                times,
                t
            )
        );
    }

    std::sort(
        exerciseIndices.begin(),
        exerciseIndices.end()
    );

    exerciseIndices.erase(
        std::unique(
            exerciseIndices.begin(),
            exerciseIndices.end()
        ),
        exerciseIndices.end()
    );
}

/*
    Pathwise discount bond P(t,T).
*/

double LSMBermudanPricer::discountBondPathwise(
    int startIndex,
    double endTime,
    int pathIndex
) const {

    int endIndex =
        nearestTimeIndex(
            times,
            endTime
        );

    if(endIndex <= startIndex){
        return 1.0;
    }

    return pathDiscountFactorBetweenIndices(
        times,
        ratePaths[pathIndex],
        startIndex,
        endIndex
    );
}

/*
    Pathwise forward swap rate and annuity.
*/

void LSMBermudanPricer::swapRateAndAnnuityPathwise(
    int exerciseIndex,
    int pathIndex,
    double& swapRate,
    double& annuity
) const {

    double exerciseTime =
        times[exerciseIndex];

    std::vector<double> payDates =
        instrument->paymentDates(
            exerciseTime
        );

    double accrual =
        1.0
        /
        static_cast<double>(
            instrument->paymentFrequency
        );

    annuity = 0.0;

    std::vector<double> dfs;

    for(double T : payDates){

        double df =
            discountBondPathwise(
                exerciseIndex,
                T,
                pathIndex
            );

        dfs.push_back(df);

        annuity +=
            accrual
            *
            df;
    }

    if(dfs.empty() || std::abs(annuity) < 1e-12){

        swapRate = 0.0;
        return;
    }

    double floatLeg =
        1.0
        -
        dfs.back();

    swapRate =
        floatLeg
        /
        annuity;
}

/*
    Immediate exercise values across all paths.
*/

std::vector<double> LSMBermudanPricer::immediateExerciseValues(
    int exerciseIndex,
    std::vector<double>& swapRates
) const {

    int nPaths =
        static_cast<int>(
            ratePaths.size()
        );

    std::vector<double> values(
        nPaths,
        0.0
    );

    swapRates =
        std::vector<double>(
            nPaths,
            0.0
        );

    for(int path = 0; path < nPaths; ++path){

        double swapRate;
        double annuity;

        swapRateAndAnnuityPathwise(
            exerciseIndex,
            path,
            swapRate,
            annuity
        );

        swapRates[path] =
            swapRate;

        double swapValue;

        if(instrument->payer){

            swapValue =
                instrument->notional
                *
                annuity
                *
                (
                    swapRate
                    -
                    instrument->fixedRate
                );

        } else {

            swapValue =
                instrument->notional
                *
                annuity
                *
                (
                    instrument->fixedRate
                    -
                    swapRate
                );
        }

        values[path] =
            std::max(
                swapValue,
                0.0
            );
    }

    return values;
}

/*
    Solve least squares by normal equations:

        beta = (X'X)^(-1) X'y

    This is simple and dependency-free.
    For production use, prefer QR decomposition.
*/

std::vector<double> LSMBermudanPricer::solveLeastSquares(
    const std::vector<std::vector<double>>& X,
    const std::vector<double>& y
) const {

    int n =
        static_cast<int>(
            X.size()
        );

    int k =
        static_cast<int>(
            X[0].size()
        );

    std::vector<std::vector<double>> A(
        k,
        std::vector<double>(
            k,
            0.0
        )
    );

    std::vector<double> b(
        k,
        0.0
    );

    for(int i = 0; i < n; ++i){

        for(int col = 0; col < k; ++col){

            b[col] +=
                X[i][col]
                *
                y[i];

            for(int row = 0; row < k; ++row){

                A[col][row] +=
                    X[i][col]
                    *
                    X[i][row];
            }
        }
    }

    /*
        Gaussian elimination with partial pivoting.
    */

    for(int i = 0; i < k; ++i){

        int pivot =
            i;

        for(int row = i + 1; row < k; ++row){

            if(std::abs(A[row][i]) > std::abs(A[pivot][i])){

                pivot =
                    row;
            }
        }

        std::swap(
            A[i],
            A[pivot]
        );

        std::swap(
            b[i],
            b[pivot]
        );

        double diag =
            A[i][i];

        if(std::abs(diag) < 1e-12){

            diag =
                1e-12;
        }

        for(int col = i; col < k; ++col){

            A[i][col] /=
                diag;
        }

        b[i] /=
            diag;

        for(int row = 0; row < k; ++row){

            if(row == i){
                continue;
            }

            double factor =
                A[row][i];

            for(int col = i; col < k; ++col){

                A[row][col] -=
                    factor
                    *
                    A[i][col];
            }

            b[row] -=
                factor
                *
                b[i];
        }
    }

    return b;
}

/*
    LSM price.
*/

double LSMBermudanPricer::price(){

    int nPaths =
        static_cast<int>(
            ratePaths.size()
        );

    int nSteps =
        static_cast<int>(
            times.size()
        );

    cashflows =
        std::vector<double>(
            nPaths,
            0.0
        );

    exerciseTimeIndex =
        std::vector<int>(
            nPaths,
            -1
        );

    exerciseFlagMatrix =
        std::vector<std::vector<bool>>(
            nPaths,
            std::vector<bool>(
                nSteps,
                false
            )
        );

    regressionDiagnostics.clear();

    std::vector<int> reversedIndices =
        exerciseIndices;

    std::reverse(
        reversedIndices.begin(),
        reversedIndices.end()
    );

    /*
        Start from final exercise date.
    */

    int lastExerciseIndex =
        reversedIndices[0];

    std::vector<double> swapRates;

    std::vector<double> immediate =
        immediateExerciseValues(
            lastExerciseIndex,
            swapRates
        );

    for(int path = 0; path < nPaths; ++path){

        cashflows[path] =
            immediate[path];

        if(immediate[path] > 0.0){

            exerciseTimeIndex[path] =
                lastExerciseIndex;

            exerciseFlagMatrix[path][lastExerciseIndex] =
                true;
        }
    }

    /*
        Move backward through exercise dates.
    */

    for(size_t idx = 1; idx < reversedIndices.size(); ++idx){

        int exerciseIndex =
            reversedIndices[idx];

        immediate =
            immediateExerciseValues(
                exerciseIndex,
                swapRates
            );

        std::vector<int> regressionPaths;

        for(int path = 0; path < nPaths; ++path){

            bool itm =
                immediate[path] > 0.0;

            bool active =
                (
                    exerciseTimeIndex[path] > exerciseIndex
                    ||
                    exerciseTimeIndex[path] == -1
                );

            if(itm && active){

                regressionPaths.push_back(
                    path
                );
            }
        }

        std::vector<double> continuation(
            nPaths,
            0.0
        );

        if(regressionPaths.size() > 5){

            std::vector<std::vector<double>> X;

            std::vector<double> y;

            for(int path : regressionPaths){

                int futureIndex =
                    exerciseTimeIndex[path];

                if(futureIndex == -1){

                    futureIndex =
                        exerciseIndices.back();
                }

                double df =
                    pathDiscountFactorBetweenIndices(
                        times,
                        ratePaths[path],
                        exerciseIndex,
                        futureIndex
                    );

                double discountedFuture =
                    cashflows[path]
                    *
                    df;

                double shortRate =
                    ratePaths[path][exerciseIndex];

                std::vector<double> basis =
                    buildBasisVector(
                        shortRate,
                        swapRates[path],
                        basisType,
                        basisDegree
                    );

                X.push_back(
                    basis
                );

                y.push_back(
                    discountedFuture
                );
            }

            std::vector<double> beta =
                solveLeastSquares(
                    X,
                    y
                );

            for(size_t k = 0; k < regressionPaths.size(); ++k){

                int path =
                    regressionPaths[k];

                double shortRate =
                    ratePaths[path][exerciseIndex];

                std::vector<double> basis =
                    buildBasisVector(
                        shortRate,
                        swapRates[path],
                        basisType,
                        basisDegree
                    );

                double fitted =
                    0.0;

                for(size_t b = 0; b < beta.size(); ++b){

                    fitted +=
                        beta[b]
                        *
                        basis[b];
                }

                continuation[path] =
                    fitted;
            }

            RegressionDiagnostic diag;

            diag.exerciseTime =
                times[exerciseIndex];

            diag.numberOfRegressionPaths =
                static_cast<int>(
                    regressionPaths.size()
                );

            diag.coefficients =
                beta;

            regressionDiagnostics.push_back(
                diag
            );
        }

        /*
            Exercise decision.
        */

        for(int path = 0; path < nPaths; ++path){

            bool itm =
                immediate[path] > 0.0;

            bool active =
                (
                    exerciseTimeIndex[path] > exerciseIndex
                    ||
                    exerciseTimeIndex[path] == -1
                );

            bool exerciseNow =
                itm
                &&
                active
                &&
                (
                    immediate[path]
                    >
                    continuation[path]
                );

            if(exerciseNow){

                cashflows[path] =
                    immediate[path];

                exerciseTimeIndex[path] =
                    exerciseIndex;

                std::fill(
                    exerciseFlagMatrix[path].begin(),
                    exerciseFlagMatrix[path].end(),
                    false
                );

                exerciseFlagMatrix[path][exerciseIndex] =
                    true;
            }
        }
    }

    /*
        Discount selected cashflows to time 0.
    */

    double sum =
        0.0;

    for(int path = 0; path < nPaths; ++path){

        int exIndex =
            exerciseTimeIndex[path];

        if(exIndex == -1){
            continue;
        }

        double df =
            pathDiscountFactorToIndex(
                times,
                ratePaths[path],
                exIndex
            );

        sum +=
            cashflows[path]
            *
            df;
    }

    return
        sum
        /
        static_cast<double>(
            nPaths
        );
}

/*
    Print exercise summary.
*/

void LSMBermudanPricer::printExerciseSummary() const {

    std::cout
        << "\nExercise Summary"
        << std::endl;

    std::cout
        << "============================================"
        << std::endl;

    for(int exIndex : exerciseIndices){

        int count =
            0;

        for(int idx : exerciseTimeIndex){

            if(idx == exIndex){
                count++;
            }
        }

        std::cout
            << "time="
            << times[exIndex]
            << ", count="
            << count
            << ", ratio="
            << static_cast<double>(count)
               /
               static_cast<double>(exerciseTimeIndex.size())
            << std::endl;
    }

    int never =
        0;

    for(int idx : exerciseTimeIndex){

        if(idx == -1){
            never++;
        }
    }

    std::cout
        << "never exercised, count="
        << never
        << ", ratio="
        << static_cast<double>(never)
           /
           static_cast<double>(exerciseTimeIndex.size())
        << std::endl;
}

/*
    Print regression diagnostics.
*/

void LSMBermudanPricer::printRegressionDiagnostics() const {

    std::cout
        << "\nRegression Diagnostics"
        << std::endl;

    std::cout
        << "============================================"
        << std::endl;

    for(const auto& diag : regressionDiagnostics){

        std::cout
            << "exercise time="
            << diag.exerciseTime
            << ", paths="
            << diag.numberOfRegressionPaths
            << ", coefficients=";

        for(double c : diag.coefficients){

            std::cout
                << c
                << " ";
        }

        std::cout
            << std::endl;
    }
}

File: 12_bermudan_swaption_lsm_pricing/python/lsm_bermudan_pricer.py
"""
lsm_bermudan_pricer.py

Longstaff-Schwartz Monte Carlo pricer for Bermudan swaptions.

The central idea:

    immediate exercise value
        vs
    estimated continuation value

The continuation value is estimated by regression on simulated paths.
"""

import numpy as np
import pandas as pd

from regression_basis import (
    polynomial_basis,
    rate_and_swap_basis
)

from lsm_path_simulation import (
    nearest_time_index,
    path_discount_factors_between_indices,
    path_discount_factors_to_index
)


class BermudanSwaptionLSMPricer:
    """
    Bermudan swaption Longstaff-Schwartz pricer.

    Parameters
    ----------
    model : HullWhiteModel
        Chapter 09 Hull-White model.

    instrument : BermudanSwaption
        Chapter 11 Bermudan swaption instrument.

    times : np.ndarray
        Simulation time grid.

    rate_paths : np.ndarray
        Simulated short-rate paths.

    basis_type : str
        "rate", "swap", or "rate_swap".

    basis_degree : int
        Polynomial basis degree.
    """

    def __init__(
        self,
        model,
        instrument,
        times,
        rate_paths,
        basis_type="rate_swap",
        basis_degree=2
    ):

        self.model = model
        self.instrument = instrument

        self.times = times
        self.rate_paths = rate_paths

        self.basis_type = basis_type
        self.basis_degree = basis_degree

        self.exercise_indices = [
            nearest_time_index(
                times,
                t
            )
            for t in instrument.exercise_dates
        ]

        self.exercise_indices = sorted(
            list(
                set(
                    self.exercise_indices
                )
            )
        )

        self.cashflows = None
        self.exercise_time_index = None
        self.exercise_flag_matrix = None

        self.regression_diagnostics = []

    def discount_bond_pathwise(
        self,
        start_index,
        end_time,
        path_index
    ):
        """
        Approximate P(t,T) pathwise using simulated short rates.

        This is different from Chapter 11 tree pricing, where P(t,T)
        came from the affine Hull-White bond pricing formula.

        In LSM, we use the simulated path between t and T.
        """

        end_index = nearest_time_index(
            self.times,
            end_time
        )

        if end_index <= start_index:
            return 1.0

        dt = self.times[1] - self.times[0]

        integral = np.sum(
            self.rate_paths[
                path_index,
                start_index:end_index
            ]
        ) * dt

        return np.exp(
            -integral
        )

    def swap_rate_and_annuity_pathwise(
        self,
        exercise_index,
        path_index
    ):
        """
        Compute pathwise forward swap rate and annuity at an exercise time.
        """

        exercise_time = self.times[
            exercise_index
        ]

        pay_dates = self.instrument.payment_dates(
            exercise_time
        )

        accrual = (
            1.0
            /
            self.instrument.payment_frequency
        )

        dfs = np.array(
            [
                self.discount_bond_pathwise(
                    start_index=exercise_index,
                    end_time=T,
                    path_index=path_index
                )
                for T in pay_dates
            ]
        )

        annuity = accrual * np.sum(
            dfs
        )

        if len(dfs) == 0:
            return 0.0, 0.0

        float_leg = (
            1.0
            -
            dfs[-1]
        )

        if abs(annuity) < 1e-12:
            return 0.0, annuity

        swap_rate = (
            float_leg
            /
            annuity
        )

        return swap_rate, annuity

    def immediate_exercise_values(
        self,
        exercise_index
    ):
        """
        Compute immediate exercise value for all paths at one exercise date.
        """

        n_paths = self.rate_paths.shape[0]

        values = np.zeros(
            n_paths
        )

        swap_rates = np.zeros(
            n_paths
        )

        annuities = np.zeros(
            n_paths
        )

        for path in range(n_paths):

            swap_rate, annuity = (
                self.swap_rate_and_annuity_pathwise(
                    exercise_index,
                    path
                )
            )

            swap_rates[path] = swap_rate
            annuities[path] = annuity

            if self.instrument.payer:

                swap_value = (
                    self.instrument.notional
                    *
                    annuity
                    *
                    (
                        swap_rate
                        -
                        self.instrument.fixed_rate
                    )
                )

            else:

                swap_value = (
                    self.instrument.notional
                    *
                    annuity
                    *
                    (
                        self.instrument.fixed_rate
                        -
                        swap_rate
                    )
                )

            values[path] = max(
                swap_value,
                0.0
            )

        return values, swap_rates, annuities

    def build_basis(
        self,
        exercise_index,
        path_indices,
        swap_rates
    ):
        """
        Build regression design matrix.
        """

        short_rates = self.rate_paths[
            path_indices,
            exercise_index
        ]

        if self.basis_type == "rate":

            return polynomial_basis(
                short_rates,
                degree=self.basis_degree
            )

        elif self.basis_type == "swap":

            return polynomial_basis(
                swap_rates[path_indices],
                degree=self.basis_degree
            )

        elif self.basis_type == "rate_swap":

            return rate_and_swap_basis(
                short_rates,
                swap_rates[path_indices],
                degree=self.basis_degree
            )

        else:

            raise ValueError(
                "basis_type must be 'rate', 'swap', or 'rate_swap'."
            )

    def price(
        self
    ):
        """
        Price Bermudan swaption using Longstaff-Schwartz Monte Carlo.

        Algorithm:
        ----------
        1. Start from final exercise date.
        2. Move backward through exercise dates.
        3. At each date, regress discounted future cashflows on state variables.
        4. Exercise if immediate value exceeds estimated continuation value.
        """

        n_paths = self.rate_paths.shape[0]

        n_steps = self.rate_paths.shape[1]

        cashflows = np.zeros(
            n_paths
        )

        exercise_time_index = np.full(
            n_paths,
            -1,
            dtype=int
        )

        exercise_flag_matrix = np.zeros(
            (
                n_paths,
                n_steps
            ),
            dtype=bool
        )

        self.regression_diagnostics = []

        exercise_indices_reversed = list(
            reversed(
                self.exercise_indices
            )
        )

        last_exercise_index = exercise_indices_reversed[0]

        immediate, swap_rates, annuities = (
            self.immediate_exercise_values(
                last_exercise_index
            )
        )

        cashflows = immediate.copy()

        exercised = immediate > 0.0

        exercise_time_index[exercised] = (
            last_exercise_index
        )

        exercise_flag_matrix[
            exercised,
            last_exercise_index
        ] = True

        for exercise_index in exercise_indices_reversed[1:]:

            immediate, swap_rates, annuities = (
                self.immediate_exercise_values(
                    exercise_index
                )
            )

            itm = immediate > 0.0

            active = (
                exercise_time_index
                >
                exercise_index
            ) | (
                exercise_time_index
                ==
                -1
            )

            regression_paths = np.where(
                itm & active
            )[0]

            continuation = np.zeros(
                n_paths
            )

            if len(regression_paths) > 5:

                future_indices = exercise_time_index[
                    regression_paths
                ]

                future_indices = np.where(
                    future_indices == -1,
                    self.exercise_indices[-1],
                    future_indices
                )

                discounted_future_values = np.zeros(
                    len(regression_paths)
                )

                for k, path in enumerate(regression_paths):

                    future_index = future_indices[k]

                    df = path_discount_factors_between_indices(
                        self.times,
                        self.rate_paths[[path], :],
                        exercise_index,
                        future_index
                    )[0]

                    discounted_future_values[k] = (
                        cashflows[path]
                        *
                        df
                    )

                X = self.build_basis(
                    exercise_index,
                    regression_paths,
                    swap_rates
                )

                beta, *_ = np.linalg.lstsq(
                    X,
                    discounted_future_values,
                    rcond=None
                )

                continuation[regression_paths] = (
                    X @ beta
                )

                self.regression_diagnostics.append(
                    {
                        "exercise_time": self.times[exercise_index],
                        "n_regression_paths": len(regression_paths),
                        "basis_type": self.basis_type,
                        "basis_degree": self.basis_degree,
                        "coefficients": beta
                    }
                )

            exercise_now = (
                itm
                &
                active
                &
                (
                    immediate > continuation
                )
            )

            cashflows[exercise_now] = immediate[
                exercise_now
            ]

            exercise_time_index[exercise_now] = (
                exercise_index
            )

            exercise_flag_matrix[
                exercise_now,
                :
            ] = False

            exercise_flag_matrix[
                exercise_now,
                exercise_index
            ] = True

        discounted_cashflows = np.zeros(
            n_paths
        )

        for path in range(n_paths):

            ex_idx = exercise_time_index[path]

            if ex_idx == -1:

                discounted_cashflows[path] = 0.0

            else:

                df = path_discount_factors_to_index(
                    self.times,
                    self.rate_paths[[path], :],
                    ex_idx
                )[0]

                discounted_cashflows[path] = (
                    cashflows[path]
                    *
                    df
                )

        self.cashflows = cashflows
        self.exercise_time_index = exercise_time_index
        self.exercise_flag_matrix = exercise_flag_matrix

        return float(
            np.mean(
                discounted_cashflows
            )
        )

    def exercise_summary(
        self
    ):
        """
        Return exercise distribution by time.
        """

        if self.exercise_time_index is None:
            raise ValueError(
                "Run price() before calling exercise_summary()."
            )

        rows = []

        for idx in self.exercise_indices:

            count = int(
                np.sum(
                    self.exercise_time_index
                    ==
                    idx
                )
            )

            rows.append(
                {
                    "time": self.times[idx],
                    "exercise_count": count,
                    "exercise_ratio": count / len(
                        self.exercise_time_index
                    )
                }
            )

        never = int(
            np.sum(
                self.exercise_time_index
                ==
                -1
            )
        )

        rows.append(
            {
                "time": np.nan,
                "exercise_count": never,
                "exercise_ratio": never / len(
                    self.exercise_time_index
                )
            }
        )

        return pd.DataFrame(rows)